daft/expressions/expressions.py

```diff
@@ -1436,6 +1436,30 @@ def substr(self, start: int | Expression, length: int | Expression | None = None
         length_expr = Expression._to_expression(length)
         return Expression._from_pyexpr(self._expr.utf8_substr(start_expr._expr, length_expr._expr))
 
+    def to_date(self, format: str | Expression) -> Expression:
+        """Converts a string to a date using the specified format
+
+        Example:
+            >>> col("x").str.to_date("%Y-%m-%d")
+
+        Returns:
+            Expression: a Date expression which is parsed by given format
+        """
+        format_expr = Expression._to_expression(format)
+        return Expression._from_pyexpr(self._expr.utf8_to_date(format_expr._expr))
+
+    def to_datetime(self, format: str | Expression) -> Expression:
```

Comment:
We need to support different timeunits, as well as datetimes with timezones. To do this, lets change the function signature to `to_datetime(self, format : str, timezone: str | None = None)`

By constraining format to be a single str, we can infer the correct timeunit when we call `to_field` in src/daft-dsl/src/functions/utf8/to_datetime.rs by checking if the format string contains nanosecond, i.e. `%9f` or `%.9f`, or millisecond, i.e. `%.3f` or `%3f`, precision, otherwise default to microsecond.

By allowing the `timezone: str | None = None` argument, we can parse the string using `chrono::NaiveDateTime::parse_from_str` or `chrono::DateTime::parse_from_str` at the array level.